Beginning of Rust support in Zephyr

d3zd3z opened the pull request:
This package implements the bare minimum support needed to build a simple Zephyr application written in Rust. There are no bindings or Zephyr API's provided, just the build infrastructure. The is described in #75900.

d3zd3z committed:
rust: Initial config option

Add the `CONFIG_RUST` Kconfig.  This can be set to indicate that an
application wishes to use Rust support.

Adds `CONFIG_RUST_SUPPORTED` to indicate what platforms Rust is known to
work on.  This is set to the targets that are supported by subsequent
commits.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Simple main wrapper

Until further variants are needed, this provides a `main()` function
that simply calls into the `rust_main()` function.  This is adequate for
applications where main can be directly written in Rust.

The reason that this is written here, rather than just writing `main`
directly in Rust is that `kernel/main_weak.c` provides a weak default
implementation that will be brought in before the Rust application's
`main` is linked in.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Basic `zephyr` crates

This is the initial Zephyr-specific crates that provide support for Rust
on Zephyr.  Currently, what they do is fairly minimal, but important.

`zephyr-build` is a build time crate (linked against the `zephyr`
crate's build.rs) that processes the current build's Kconfig settings,
and does three things with them:

  - Boolean Kconfig values given given to the Rust toolchain so that
    conditional compilation can be based off of them.
  - Numeric Kconfig settings end up as constants in `zephyr::kconfig`.
  - String valued Kconfig settings end up as constants in
    `zephyr::kconfig`.

None of these cause code or data to be generated or allocated (but note
that if there is a reference to a string, that string will be placed in
read-only memory).

The `zephyr` crate is built for the target, and intended to be
referenced by the application.  It provides minimal support needed for a
"bare" Rust build, notably a panic handler.  At this point, the panic
handler is not implemented as we need better support for calling into
Zephyr's C code, so it just stops in an infinite loop.

It also ensures that `CONFIG_RUST` is set as a sanity check, and
includes the generated kconfig.rs file with the kconfig definitions for
the current build.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Add cmake support for building rust apps

This provides the function `rust_cargo_application()` within Cmake to
give applications written in Rust a simple way to start.

This implements the basic functionality needed to build a rust
application on Arm, including target mapping, and invoking cargo during
the build.

Most of the functionality is about passing the appropriate flags to
`cargo`, which is used to perform the actual build of the rust code.
Cargo generates `librustapp.a` which is added to the link dependencies
for the application.

The cargo rule is written such that cargo will always be built (provided
`app` is being built), as there will be complex dependencies on the
cargo side.  Cargo will not modify the `librustapp.a` file if nothing
needs to be build, so this will generally be fast if there are no
changes to files that affect the Rust build.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Simple rust hello_world application

Create the Rust equivalent of the hello_world application.  This shows
the use of printk from Rust as well as accessing a string value from
Kconfig.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
MAINTAINERS: Add Rust to maintainers file

Indicate the new code is maintained.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
doc: languages: Add Rust documentation

Add initial docs for Rust language support.  Explains the support
currently present, how to install the toolchain needed, and explains
what is in the current sample.

Signed-off-by: David Brown <david.brown@linaro.org>

pdgendt commented on lib/rust/Kconfig:
```diff
@@ -0,0 +1,13 @@
+
+menu "Rust Language Support"
+
+config RUST
+	bool "Rust support for the application"
+	help
```
I think `select EXPERIMENTAL` should be added :)

d3zd3z replied:
Done

pdgendt commented on lib/rust/zephyr-build/src/lib.rs:
```diff
@@ -0,0 +1,72 @@
+        }
+        if let Some(caps) = config_str.captures(&line) {
+            writeln!(&mut f, "    #[allow(dead_code)]").unwrap();
+            writeln!(&mut f, "    pub const {}: &'static str = {};",
+                &caps[1], &caps[2]).unwrap();
```
Isn't it possible to append the `'\0'` byte here so you don't need a copy to pass the string back to the C world?

d3zd3z replied:
I'm not aware of any way to do this in stable Rust.

pdgendt replied:
I think there is support for [C-string literals](https://blog.rust-lang.org/2024/03/21/Rust-1.77.0.html#c-string-literals) since 1.77.

d3zd3z replied:
But, this isn't a C-string literal, it is a generated regular string literal that I want to use. I don't want all of the generated literals to be C-strings.

pdgendt commented on lib/rust/zephyr-build/src/lib.rs:
```diff
@@ -0,0 +1,72 @@
+        let line = line.expect("reading line from dotconfig");
+        if let Some(caps) = config_num.captures(&line) {
+            writeln!(&mut f, "    #[allow(dead_code)]").unwrap();
+            writeln!(&mut f, "    pub const {}: usize = {};",
+                &caps[1], &caps[2]).unwrap();
```
The options can contain negative numbers, I think.

d3zd3z replied:
I've changed 'int' kconfig fields to be `isize` instead of `usize`, and allow the negative values.

alwa-nordic commented:
Thank you for making this PR. It's great to get the ball rolling! Let's make Rust in Zephyr great!

But, I think this PR tries to do a bit too much too quickly. I think the most important first step is to ensure building and linking is done correctly. I think would be best to focus on this and de-scope other parts of this PR for later. Please read on below where I try to explain.

## LLVM triplets

This PR is using LLVM for compiling Rust, which I think is a good idea. But it does mean we need to think about ABI compatibility with the Zephyr toolchain.

Linking with LLVM objects is not a Rust-specific problem, and we should solve this first, possibly in a language-agnostic way. We should start with a solid foundation on how to correctly build LLVM objects for linking with the Zephyr ABIs.

I think the as first thing on the roadmap to Rust, Zephyr should have an utility that selects a foolproof LLVM target specification. And we must have tests that can verify that it works at runtime, e.g. by calling various types of functions and checking for data corruption.

The reason this may be a concern is that Zephyr defines its own ABI, which may or may not be equivalent to any of the well-known LLVM triplets. For example, in Zephyr `int` is always 32 bits, even on 64-bit platforms. Without a test suite, it's too easy to mess up here.

The PR uses the well-known `thumbv8m.main-none-eabi` and others. We need tests to confirm that we are not making a mistake here.

For prior art, see [tylerwhall/zephyr-rust target specifications](https://github.com/tylerwhall/zephyr-rust/tree/master/rust/targets), which defines new LLVM triplets for Zephyr OS. This project also Apache-licensed.

## Kconfig in Rust

I feel a framework for exposing Kconfig values in Rust should be done at a later stage, since there are many ways to do this and it has attracted comments on how to map the Kconfig values onto Rust types. I think it's better to defer this work, so that the initial Rust support is not delayed by this.

Rust already has `core::env!` for accessing environment variables at build time. I believe this is a good-enough solution for passing in Kconfig variables for now. It's simple enough to copy from Kconfig variables to environment variables in CMakeLists.txt.

I also worry there are good reasons for why `core::env!` is not like this PR, generating a module with symbols for all environment variables. Of course, it's fine to experiment.

## Safety in samples

The sample declares `printk` as a two-argument function. But `printk` is actually a variadic C function. This not the same type so it is probably unsound.

Safety is integral to Rust culture. I think that all Rust samples code should be `#![forbid(unsafe_code)]` to reflect that. That means FFI calls in samples must be trough safe wrappers.

d3zd3z commented:
> But, I think this PR tries to do a bit too much too quickly.

So, I think we have different philosophies about what we need to do to get rust support into Zephyr. Yes, there is stuff that we need to do "right", but if we are too careful, we'll never get anything at all. My goal here is just to go through the build stuff, and I have a _lot_ of other stuff in process that needs to builds to be able to happen. I guess I'd rather have it work well in a limited situation than solve everything before anything can go in.

> triplets...

So, getting this right is important, but this is also difficult to come up with tests for. One possibility is to just be more restrictive about it, and pick configurations that we know work.

But, I don't think the calling is as difficult of an issue as you are imagining. Generally, getting the triple wrong is more likely to result in either code that doesn't run at all (using the wrong CPU), or fails at link time (usually from hard float mismatch).

Matching the Zephyr types is important, but I wouldn't think of that as a compiler or toolchain, but a matching the ABI. It is also not addresed at all yet, as we aren't really calling anything.

> Kconfig

I disagree about delaying it. I'm open for other ideas on how to do it, but getting access to them is foundational for most later work. Do you have specific concerns about how I'm doing it?

I'm not quite sure I understand how `core::env()` is useful. It can't be used for conditional compilation, which is one of the primary needs for getting to bool Kconfig values.

> samples

As far as the samples, it isn't intended to stay that way, and in fact, my other PR has already removed it.

Basically, I'm concerned about spinning, while bike shedding things that really aren't that important. One of my goals is to make using Rust inside of Zephyr as nice (well, hopefully nicer) as writing C code. Coming next is making the DT available to Rust code.

d3zd3z commented:
A little more on my philosophy here. There are probably 30-40 things that need to be in place for Rust on Zephyr to even be slightly useful. If we sit around and try to come up with the "perfect" solution to all of these problems, I don't think we'll ever get to the point where there is something useful. These pieces can certainly be improved once we get there. Maybe some of this could be worked on in parallel. But, not being able to even build something basic will keep us from ever getting past the first few things.

In other words, I'd like to keep things that are experimental away from affecting the rest of Zephyr. Perhaps as we really figure out what is needed, some things like target selection can be improved on, but I'd really like to get to the point where someone could reasonably write an actual application in Rust that runs on Zephyr. That means it needs kconfig, it needs to see the device tree, and have accessors into the device tree to give it save wrappers to be able to use real Zephyr devices. Threads, syscalls, synchronization, logging, etc all needs to be there before it's really useful. I think we are much better getting the pieces in place, adequately, not perfectly, will give this a much better chance of success. Once it is actually useful, the pieces can be improved and perfected. This is, in fact, how the rest of Zephyr's development has happened.

I do agree that, for example, applications shouldn't need unsafe. That was just a temporary placeholder until there is enough functionality to have something to safely call. It is surprising how much it takes to get there.

cfriedt commented:
I don't think we need to get too hung up on LLVM triplets; if a triplet is incorrect, there will be linker errors, so there is not a lot of wiggle room anyway.

Kconfig is kind of special because they are both macros and absolute symbols in Zephyr. For Rust, I would imagine that we would just generate similar macros and that we would leave the absolute symbols as they are, as part of the C build.

Devicetree will be interesting, I think.

David has already covered much of this in the RFC and his previous talks on the subject though as well.

I would say it's best to get to something functional first, and then tweak things later. The more we can have tested in CI, the better, so that we know assumptions are sound.

Pixep commented on cmake/modules/rust.cmake:
```diff
@@ -0,0 +1,145 @@
+# Rust make support
+
+# Zephyr targets are defined through Kconfig.  We need to map these to
+# an appropriate llbm target triple.  This sets `RUST_TARGET` in the
```
```suggestion
# an appropriate llvm target triple.  This sets `RUST_TARGET` in the
```
Little typo

Ayush1325 commented on doc/develop/languages/rust/index.rst:
What is the reason for using a C file to call rust main? Is it related to wanting to keep the signature similar to the Rust std main?

d3zd3z replied:
There isn't a real strong reason to do this. The main issue I ran into is that Zephyr already has a weak symbol for 'main', if it is provided by a library instead of a C source, the linker prefers the weak symbol, and the rust code never gets called.

Ayush1325 commented on doc/develop/languages/rust/index.rst:
I am not sure how I feel about using `build.rs` here. While it is fine, this is a cargo specific thing that any other build system with Rust support doesn't really like.
While there are a lot of crates that use it, I would prefer to avoid it if possible. Have you looked at the Linux kernel implementation for Kconfig in Rust?

d3zd3z replied:
I _really_ want to make rust applications fully able to work with Cargo. Maybe there are other situations (such as adding code within Zephyr in Rust) where calling rustc directly would be useful. But for an application, where someone is quite likely to want external dependencies anyway, trying to do everything that cargo does is needlessly complicated.

Also, this is an almost trivial example of the kind of stuff we will need at build time to make Rust useful within Zephyr. In my other PR that brings in device tree, I'm also generating a devicetree.rs from the device tree for a particular build. It needs to run at build time.

I'm kind of trying to understand where the fear of cargo is coming from. Along with the memory issues and such, cargo and the crates community is one of the big reasons for someone to want to use Rust.

What I'm doing here is _very_ different from how Rust is being used in Linux. This is not about extending Zephyr with Rust, but about writing applications. It's more akin to writing Rust programs in Linux userspace.

Ayush1325 replied:
The problem with `build.rs` is that it is cargo specific rather than Rust specific. There are some other build systems that support Rust [(meson)](https://mesonbuild.com/), but they do not support `build.rs`.

teburd replied:
I’m not sure what you are looking for here, supporting application development with rust cargo makes immense sense.

Ayush1325 replied:
Yes, it makes sense to use cargo for application development, but by tying the devicetree and other buildtime stuff that Rust support needs to `build.rs` might have some unintended consequences:
1. The current devicetree implementation will not be compatible with writing drivers and zephyr internals which I imagine would not go the cargo route.
2. Make it impossible to use other build systems for applications since they do not support `build.rs`.

Note: I do not have any problem with running a rust file at build time. I just do not want to tie the build time generation we need to do to cargo.

d3zd3z replied:
I'm not exactly sure what you would propose. `build.rs` is _how_ you do build-time extensions with Cargo. The file is very short, and calls code in `zephyr-build`. The bindgen code is more complex, but still can be handled whever.

I don't want to actively go against people that want to build Rust apps without using cargo, but I'm still a bit baffled as to why someone would want to do that. Cargo does a lot of stuff for us, and for the app developer, and it would be rather silly to try and reproduce all of that in Cmake. If Meson has replicated some of that work, then there would be a need to port that effort to Meson.

We will have many additional issues to address to incorporate rust code into Zephyr rather than just to use it, the least of which is getting a lot more broad approval of Rust.

Ayush1325 reviewed (commented):
Overall, looks like a good start. I am personally more interested in Rust outside cargo in Zephyr, but at least for applications, cargo is definitely useful.

Pixep commented on cmake/modules/rust.cmake:
```diff
@@ -0,0 +1,145 @@
+# Rust make support
+
+# Zephyr targets are defined through Kconfig.  We need to map these to
+# an appropriate llbm target triple.  This sets `RUST_TARGET` in the
```
```suggestion
# an appropriate llvm target triple. This sets `RUST_TARGET` in the
```

d3zd3z replied:
Applied

Pixep commented on lib/rust/zephyr-build/src/lib.rs:
```diff
@@ -0,0 +1,79 @@
+            writeln!(&mut f, "    #[allow(dead_code)]").unwrap();
+            writeln!(&mut f, "    pub const {}: usize = {};",
+                &caps[1], &caps[2]).unwrap();
+        }
+        if let Some(caps) = config_int.captures(&line) {
+            writeln!(&mut f, "    #[allow(dead_code)]").unwrap();
+            writeln!(&mut f, "    pub const {}: isize = {};",
+                &caps[1], &caps[2]).unwrap();
+        }
+        if let Some(caps) = config_str.captures(&line) {
```
```suggestion
        } else if let Some(caps) = config_int.captures(&line) {
            writeln!(&mut f, "    #[allow(dead_code)]").unwrap();
            writeln!(&mut f, "    pub const {}: isize = {};",
                &caps[1], &caps[2]).unwrap();
        } else if let Some(caps) = config_str.captures(&line) {
```
This would clarify that those are mutually exclusive

d3zd3z replied:
Applied

Pixep reviewed (commented):
Tested successfully on an ESP32C3 with minor changes, the demo worked great! 👍 

One limitation limitation, however, that wasn't immediately clear to me looking at it and at https://github.com/zephyrproject-rtos/zephyr/issues/75900 , is that all `static inline` function (such as `k_msleep()`) get optimized out, and fail to resolve by the linker.

The (future) fix is either to wrap symbols that need to be exported in functions or to drop `static inline` for a selected function when building a Rust app. Neither feel particularily good without actively finding symbols used and dynamically generating code to export those.

jhedberg commented:
Thanks for the excellent work!

I was wondering why I had a leftover `samples/rust/hello_rust` directory after I briefly tried this out (in a separate local git branch), and the reason seems to be `Cargo.lock` that's generated and left in the app directory:
```
❯ git status
On branch pr/75904
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	samples/rust/hello_rust/Cargo.lock
```

Ideally I suppose any generated files should only go into the Zephyr build directory, however in this case I suspect that's not feasible. I'd therefore suggest to add as part of this PR an entry for `Cargo.lock` to `.gitignore`.

alwa-nordic commented:
> Thanks for the excellent work!
> 
> I was wondering why I had a leftover `samples/rust/hello_rust` directory after I briefly tried this out (in a separate local git branch), and the reason seems to be `Cargo.lock` that's generated and left in the app directory:
> 
> ```
> ❯ git status
> On branch pr/75904
> Untracked files:
>   (use "git add <file>..." to include in what will be committed)
> 	samples/rust/hello_rust/Cargo.lock
> ```
> 
> Ideally I suppose any generated files should only go into the Zephyr build directory, however in this case I suspect that's not feasible. I'd therefore suggest to add (as part of this PR) `Cargo.lock` to `.gitignore`.

[Rust recommends that `Cargo.lock` should be checked into git.](https://doc.rust-lang.org/cargo/faq.html#why-have-cargolock-in-version-control)

d3zd3z commented on cmake/modules/rust.cmake:
I'm not exactly sure what would be upstreamed here. _everything_ in this cmake file is specific to building this within Zephyr. I don't see how it would be useful in any other context.

d3zd3z commented:
> One limitation limitation, however, that wasn't immediately clear to me looking at it and at #75900 , is that all `static inline` function (such as `k_msleep()`) get optimized out, and fail to resolve by the linker.

So solving this is a much deeper part of this work that is presented at this level. There are a couple of possible approaches. One is, as you say, to basically drop down some code that will get the C compiler to emit linked versions of the inline functions. Part of the necessity of this is that _every_ use of `static inline` is actually wrong. According to the gcc docs, it should be `extern inline`, and then there should be one instance where the code is brought in.

This could be done with some simple wrappers of C functions that just call the inline function. Since it is inlined, it will be expanded, and the definition would be basically what the inline function has in it.

The disadvantage here is that the inline doesn't happen when rust makes use of this code. So the code is a bit slower when running in Rust.

The other possibility is to actually implement all of the inline functions in Rust as well. Given the large number of these, it is probably a bit of an overwhelming maintenance nightmare, and automatically converting inline C declarations into Rust is not an easy problem. There are also lots of things that are implemented as macros that really are just code.

I'm open to ideas on how to best handle this. For some of the stuff I'm playing with in #76074, I've written the code in Rust (it doesn't add any C code, such as wrappers). Please take a look at that if you have any comments about the direction I'm taking.

d3zd3z commented:
> however in this case I suspect that's not feasible. I'd therefore suggest to add (as part of this PR) `Cargo.lock` to `.gitignore`

I think that is the best way to handle this. I don't think it makes a lot of sense to commit the Cargo.lock file, even though these are technically applications.

d3zd3z committed:
lib: rust: Add simple printk support to Rust

When `CONFIG_PRINTK` is enabled, provide macros `printk` and `printkln`
within the `zephyr` crate that applications can use to print.  This
currently sends messages, with a small amount of buffering, using
`k_str_out`.

Until C functions are made generally available from the Rust side,
we use a small wrapper function to call `k_str_out`, as it is
implemented as an inline syscall.

Signed-off-by: David Brown <david.brown@linaro.org>

mjaun commented on cmake/modules/rust.cmake:
```diff
@@ -0,0 +1,155 @@
+      set(RUST_TARGET "thumbv6m-none-eabi" PARENT_SCOPE)
+    elseif(CONFIG_CPU_CORTEX_M3)
+      set(RUST_TARGET "thumbv7m-none-eabi" PARENT_SCOPE)
+    elseif(CONFIG_CPU_CORTEX_M4)
```
```suggestion
    elseif(CONFIG_CPU_CORTEX_M4 OR CONFIG_CPU_CORTEX_M7)
```

d3zd3z replied:
I've added a patch for this.

d3zd3z committed:
rust: Ignore unknown cfgs in the zephyr crate

We use the `#[cfg(...)]` directive in rust to pass Kconfig values through.
Because it is possible for the code to depend on a Kconfig value that isn't
present in a given build, there isn't a way for us to tell the Rust
toolchain about all possible config values.

For now, just disable this warning entirely.  This functionality could be
supported by the patch validation scripts, which seems like a better place
than trying to gather a list of all possible configs at build time.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
cmake: rust: Add Cortex M7 support

The Cortex-M7 is build the same as M4.  Catch it in the same rules.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z commented:
I have removed the DNM tag, and would like to consider this for merge once the CI issues are resolved and there are sufficient reviews.

d3zd3z committed:
ci: tags: Add rust tag

The rust tag associates rust tests and samples with code that adds rust
support to Zephyr.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Tag the rust sample as rust

Add the 'rust' tag so that CI will invoke this test when rust support
files have changed.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z commented:
Now that we have a collab-rust branch, and the CI changes have merged, let's let this run through CI (which will hopefully still pass), and we can do reviews here.

kartben commented on samples/rust/.gitignore:
wouldn't it make sense to move this to root .gitgnore instead?

kartben commented on samples/rust/hello_rust/prj.conf:
please add a README.rst for this sample

kartben commented on samples/rust/hello_rust/prj.conf:
please name the folder/sample `hello_world` (see ex. how we have `samples/cpp/hello_world` and of course rename everything else accordingly, ex. test case identifier etc.

d3zd3z replied:
I got a doc build failure when it had the same name as an existing test. I’m not sure why that works for c++.

kartben replied:
A doc failure sounds really weird especially as you don't have a README.rst for the sample just yet:) I would try again

kartben commented on doc/develop/languages/rust/index.rst:
```suggestion
.. code-block:: console
```

kartben commented on doc/develop/languages/rust/index.rst:
```suggestion
Currently, Zephyr supports applications written in Rust and C.  To
```

kartben replied:
btw it supports C++ too? so first sentence should maybe be clarified

kartben replied:
this typo is still there

kartben commented on doc/develop/languages/rust/index.rst:
please don't wrap lines so aggressively (72 characters? :)) -- you may wrap at 100 like other files in zephyr

kartben commented on doc/develop/languages/rust/index.rst:
```suggestion
   = help: consider downloading the target with `rustup target add thumbv7m-none-eabi`
```

kartben commented on doc/develop/languages/rust/index.rst:
```suggestion
The application should contain a :file:`CMakeLists.txt`, similar to
```

kartben commented on doc/develop/languages/rust/index.rst:
```suggestion
The application source itself should live in :file:`src/lib.rs`.  A few
```

kartben commented on doc/develop/languages/rust/index.rst:
```suggestion
   #[cfg(not(CONFIG_SCHED_DUMB))]
```

kartben reviewed (changes requested):
some documentation related feedback. I might have more when CI has a preview site up again
Thanks <3

d3zd3z committed:
rust: Move ignores to top .gitignore

Apply the rules to ignore some cargo generated or used files to the
whole tree, not just the samples.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Add README for rust hello world

Create a Readme for this sample, explaining how to build and what it
does.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Rename hello world

Rename the sample to just hello world.  As long as the actual sample
name is different, there won't be index conflicts in the documentation.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
doc: languagues: rust: Reformat

Refill the text to 100 columns instead of 72 to match the rest of the
project.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
doc: languages: rust: typo fixes

Fix various minor typos in the Rust documentation.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Include rust samples in docs

Include the Sample Rust applications in the doc generation.

Signed-off-by: David Brown <david.brown@linaro.org>

cfriedt reviewed (dismissed):
LGTM

stephanosio commented on .github/workflows/rust.yaml:
```diff
@@ -0,0 +1,178 @@
+
+          echo "ZEPHYR_SDK_INSTALL_DIR=/opt/toolchains/zephyr-sdk-$( cat SDK_VERSION )" >> $GITHUB_ENV
+
+      - name: Install Rust Toolchain
+        run: |
+          curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
+          # Add to the local environment to be able to use.
+          source "$HOME/.cargo/env"
+          # Install support for the targets that are supported.
+          for target in $`cat lib/rust/targets.txt`; do rustup target install $$target; done
```
Is this necessary if we use the Rust toolchain integrated in the CI Docker image?

stephanosio commented on .github/workflows/rust.yaml:
```diff
@@ -0,0 +1,178 @@
+name: Build Rust tests
```
Why do we need a separate workflow for building and running Rust applications? It would be better to integrate this into the existing twister workflow.

d3zd3z replied:
This is a mistake, I will remove this file entirely.

kartben reviewed (approved):
+1 for docs/samples - thanks for this @d3zd3z!
